src/lava/lib/dl/netx/utils.py

```diff
@@ -0,0 +1,173 @@
+                value = value[()]
+            return value.decode('ascii')
+        elif key in self.copy_keys:
+            return self.f[key][()].astype(int).copy()
+        elif key in self.array_keys:
+            return self.f[key][()]
+        elif isinstance(key, int) and f'{key}' in self.f.keys():
+            return NetDict(f=self.f[f'{key}'])
+        else:
+            return NetDict(f=self.f[key])
+
+    def __setitem__(self, key: str) -> None:
+        raise NotImplementedError('Set feature is not implemented.')
+
+
+def optimize_weight_bits(weight: np.ndarray) -> Tuple[
+    np.ndarray, int, int, SYNAPSE_SIGN_MODE
+]:
+    """Optimizes the weight matrix to best fit in Loihi's synapse.
+
+    Parameters
+    ----------
+    weight : np.ndarray
+        standard 8 bit signed weight matrix.
+
+    Returns
+    -------
+    np.ndarray
+        optimized weight matrix
+    int
+        weight bits
+    int
+        weight_exponent
+    SYNAPSE_SIGN_MODE
+        synapse sign mode
+    """
+    max_weight = np.max(weight)
+    min_weight = np.min(weight)
+
+    if max_weight < 0:
+        sign_mode = SYNAPSE_SIGN_MODE.INHIBITORY
+        is_signed = 0
+    elif min_weight >= 0:
+        sign_mode = SYNAPSE_SIGN_MODE.EXCITATORY
+        is_signed = 0
+    else:
+        sign_mode = SYNAPSE_SIGN_MODE.MIXED
+        is_signed = 1
+
+    if sign_mode == SYNAPSE_SIGN_MODE.MIXED:
+        pos_scale = 127 / max_weight
+        neg_scale = -128 / min_weight
+        scale = np.min([pos_scale, neg_scale])
+    elif sign_mode == SYNAPSE_SIGN_MODE.INHIBITORY:
+        scale = -256 / min_weight
+    elif sign_mode == SYNAPSE_SIGN_MODE.EXCITATORY:
+        scale = 255 / max_weight
+
+    scale_bits = np.floor(np.log2(scale)) + is_signed
+
+    precision_found = False
+    n = 8
+    while (precision_found is False) and (n > 0):
+        roundingError = np.sum(
+            np.abs(weight / (2**n) - np.round(weight / (2**n)))
+        )
+        if roundingError == 0:
+            precision_found = True
+        else:
+            n -= 1
+
+    n -= is_signed
+
+    num_weight_bits = 8 - scale_bits - n
+    weight_exponent = -scale_bits
+
+    weight = np.left_shift(weight.astype(np.int32), int(scale_bits))
+
+    return (
+        weight.astype(int),
+        int(num_weight_bits),
+        int(weight_exponent),
+        sign_mode
+    )
+
+
+def num_delay_bits(delays: np.ndarray) -> int:
+    """Calculates the number of delay bits required.
+
+    Parameters
+    ----------
+    delays : np.ndarray
+        delay vector
+
+    Returns
+    -------
+    int
+        number of delay bits.
+    """
+    if delays.min() < 0:
+        raise ValueError(
+            f'Negative delay encountered. '
+            f'Found {delays.min()=}.'
+        )
+    if delays.max() >= 63:
+        raise ValueError(
+            f'Max delay exceeded limit of 62. '
+            f'Found {delays.max()=}.'
+        )
+    return np.ceil(np.log2(delays.max())).astype(int)
```

Comment:
Is this correct? I probably have misunderstood something, but I thought in the nxsdk docs it says that if I assign a delay width of 2(?) bits, then I can use values 1 and 62, with 3 bits I get 1, 31, 62, or something like that... essentially that it does a bit shift to give a full range even at reduced precision. If that's the case, then don't we need to basically do an error-based method similar to weights.

Reply:
No that only happens to the weights. Actually that behavior is gone in Lohi 2 weights as well. The optimize weights needs reinterpretation for Loihi 2 NxModel implementation.